clamp/clamp.tex

```diff
@@ -0,0 +1,252 @@
+\newcommand{\function}{\texttt{function}}
+\newcommand{\float}{\texttt{float }}
+\newcommand{\questionc}[1]{\textcolor{red}{\textbf{Question:} #1}}
+
+\newcommand{\silvia}[1]{{ {\color{blue}{(silvia)~#1}}}}
+\newcommand{\grace}[1]{{ {\color{purple}{(grace)~#1}}}}
+\newcommand{\connor}[1]{{ {\color{teal}{(connor)~#1}}}}
+\newcommand{\mike}[1]{{ {\color{green}{(mike)~#1}}}}
+\newcommand{\todo}{{\textcolor{red}{TODO }}}
+
+
+\title{Privacy Proofs for OpenDP: Clamping}
+\author{S\'ilvia Casacuberta}
+\date{Summer 2021}
+
+\begin{document}
+
+\maketitle
+
+\tableofcontents
+
+\section{Algorithm Implementation}
+
+\subsection{Code in Rust}
+The current OpenDP library contains the \texttt{make\_clamp\_vec} function implementing the clamping function. This is defined in lines 25-38 of the file \texttt{manipulation.rs} in the Git repository\footnote{As of June 16, 2021. Since then, the code has been updated to include a more general clampable domain, which is not yet finished.} (\url{https://github.com/opendp/opendp/blob/58feb788ec78ce739caaf3cad8471c79fd5e7132/rust/opendp/src/trans/manipulation.rs#L25-L38}).
+
+\begin{figure}[ht]
+    \includegraphics[width=16cm]{clamp_rust.png}
+    \centering
+    \label{fig:code}
+\end{figure}
+
+\silvia{Flag: will change to the new clampable domain code once total ordering has been implemented.}
+
+\subsection{Pseudocode in Python}\label{sec:pseudocode}
+We present a simplified Python-like pseudocode of the Rust implementation below. The necessary definitions for the pseudocode can be found at \href{https://www.overleaf.com/project/60d215bf90b337ac02200a99}{``List of definitions used in the pseudocode"}. 
+
+\subsubsection*{Preconditions}
+To ensure the correctness of the output, we require the following preconditions:
+
+\begin{itemize}
+    \item \textbf{User-specified types:}
+    \begin{itemize}
+        \item Type \texttt{T} must have trait \texttt{TotalOrd}.\footnote{For now, the OpenDP library only implements \texttt{PartialOrd}, but \texttt{TotalOrd} will soon be implemented.}
+    \end{itemize}
+\end{itemize}
+
+\subsubsection*{Postconditions}
+\begin{itemize}
+    \item Either a valid \texttt{Transformation} is returned or an error is returned.
+\end{itemize}
+
+\begin{lstlisting}[language=Python, escapechar=|] 
+def MakeClamp(L: T, U: T): |\label{line:def}|
+    input_domain = VectorDomain(AllDomain(T))
+    output_domain = VectorDomain(IntervalDomain(L, U)) |\label{line:output}|
+    input_metric = SymmetricDistance()
+    output_metric = SymmetricDistance()
+    
+    def Relation(d_in: u32, d_out: u32) -> bool: |\label{line:rel}|
+        return d_out >= d_in*1
+    
+    def function(data: Vec(T)) -> Vec(T): |\label{line:fn}|
+        def clamp(x: T) -> T: |\label{line:clamp}|
+            return max(min(x, U), L)
+        return list(map(clamp, data)) |\label{line:map}|
+    
+    return Transformation(input_domain, output_domain, function, input_metric, output_metric, stability_relation = Relation)
```

Comment:
Why the special naming "Relation" instead of "stability_relation"? The code makes sense. Are you guys planning on switching type hints to the [type hint syntax](https://www.python.org/dev/peps/pep-0484/)? I prefer the typing syntax for hints as it looks less like a function invocation. It's a pretty minor pick, basically `Vec(T)` -> `Vec[T]`. Also, are we keeping `Vec` from rust or using python's `list` in the pseudocode? The iterator collection uses list, but the type is a `Vec`. Maybe change one or the other to be consistent.

Reply:
Regarding the special naming of "Relation": I mainly did this to save space in Section 2. Proof when proving the stability relation, so that only "Relation" appears in the text instaed of "stability_relation". And also so that the theorem statement for the stability/privacy guarantee looks the same for measurements, where we would have `privacy_relation = Relation`. But if you think this is not enough to change it I can add `stability_relation` back, no problem.

Reply:
About the type hint syntax, are you saying that the correct function type signing for clamp is `Vec(T) -> Vec[T]`? How so? So initially we had `Vec(T) -> Vec(T)` but you said that this was not correct, and so we introduced the rule "Parentheses, `()`, are used to create an instance of a domain. Square brackets, `[]`, are used to describe the type of a domain. For example, `AllDomain(i8)` is the domain of all values of type `i8.` However, the type of `AllDomain(i8)` -- the domain itself, not the elements of the domain -- is `AllDomain[i8];` note the square brackets. This typing style is inspired by the notation used in Python; see \url{https://docs.python.org/3/library/typing.html}." (this is now included in the pseudocode definitions).

Reply:
Regarding Vec vs list: so I thought we wanted the pseudocode to look as Python as possible, which is why I wrote it in list format. But then it seemed not correct to change the input and output domains, because the `VectorDomain` formulation is important and appears everywhere, and also for showing correct output domain. When you say you would change either or the other are you proposing to either: 1. Change line 148 to `return Vec(map(clamp, data))` and keep the domains as they are, or 2. To keep line 148 as it is and change the domains to something different? Approach 2 seems worse I think, because that is inventing new domains that don't exist in the Rust library. And could we not just say that `list` has type `Vec` in the list of definitions to resolve this? But perhaps approach 1 is better then, although less Python

Reply:
1. I think the part throwing me is that camel-case code in python usually represents classes. I have a soft preference towards `relation` or `stability_relation`.
2. No, the arrow in my statement was meant to communicate "replace Vec(T) with Vec[T]". The current syntax in the pseudocode was `def function(data: Vec(T)) -> Vec(T):`. The revised syntax would be `def function(data: Vec[T]) -> Vec[T]:`. 

3. You're right, there's no win here. I think we leave Vec/list inconsistent unless someone else has strong feelings.

Reply:
1. That makes sense, I had not considered the implications of having it capitzalied. I will change all of them to `relation` then.
2. Ah okay, I thought the TeX had already been changed to `Vec(T)`, but it is fixed now.
3. Okay!